File: python/33-Search-in-Rotated-Sorted-Array.py
from typing import List


# region 2 Pass Binary Search
# Time O(logn)
# Space O(1)
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # find the pivot
        left, right = 0, len(nums) - 1

        while left < right:
            middle = left + (right - left) // 2

            if nums[right] < nums[middle]:
                left = middle + 1
            else:
                right = middle
        pivot = left

        print(pivot)

        # determine if target is on left or right of pivot & search for it

        if target > nums[-1]:
            left, right = 0, pivot - 1
        else:
            left, right = pivot, len(nums) - 1
        print(left)

        while left < right:
            middle = left + (right - left) // 2

            if nums[middle] < target:
                left = middle + 1
            else:
                right = middle
        print(left)

        return left if nums[left] == target else -1


# endregion

File: c#/33-Search-in-Rotated-Sorted-Array.cs
public class Solution {
    public int Search(int[] nums, int target) {
        int l = 0, r = nums.Length - 1;

        while (l <= r) {
            int mid = l + (r - l) / 2;

            if (nums[mid] == target) return mid;

            // Step 1: Check which side is sorted
            if (nums[l] <= nums[mid]) {
                // Left side is sorted
                if (nums[l] <= target && target < nums[mid]) {
                    // target is left
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            } else {
                // Right side is sorted
                if (nums[mid] < target && target <= nums[r]) {
                    // target is right
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }
        }

        return -1;
    }
}
